tests/phpunit/tests/admin/wpUpgrader.php

```diff
@@ -0,0 +1,1430 @@
+		$this->instance->generic_strings();
+
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+
+		if ( ! defined( 'FS_CHMOD_DIR' ) ) {
+			define( 'FS_CHMOD_DIR', 0755 );
+		}
+
+		if ( ! defined( 'FS_CHMOD_FILE' ) ) {
+			define( 'FS_CHMOD_FILE', 0644 );
+		}
+
+		$destination = DIR_TESTDATA . '/upgrade/';
+		$dirlist     = array(
+			'file1.php' => array(
+				'name' => 'file1.php',
+				'type' => 'f',
+			),
+			'subdir'    => array(
+				'name' => 'subdir',
+				'type' => 'd',
+			),
+		);
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( $dirlist );
+
+		$unwritable_checks = array(
+			array( $destination . 'file1.php' ),
+			array( $destination . 'file1.php' ),
+			array( $destination . 'subdir' ),
+			array( $destination . 'subdir' ),
+		);
+
+		$this->wp_filesystem_mock
+				->expects( $this->exactly( 4 ) )
+				->method( 'is_writable' )
+				->withConsecutive( ...$unwritable_checks )
+				->willReturn( false );
+
+		$actual = $this->instance->clear_destination( $destination );
+
+		$this->assertWPError(
+			$actual,
+			'WP_Upgrader::clear_destination() did not return a WP_Error object'
+		);
+
+		$this->assertSame(
+			'files_not_writable',
+			$actual->get_error_code(),
+			'The WP_Error code was not "files_not_writable"'
+		);
+
+		$this->assertSameSets(
+			array( 'file1.php, subdir' ),
+			$actual->get_all_error_data(),
+			'The WP_Error object did not contain the expected data'
+		);
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::install_package()` returns a WP_Error object
+	 * when invalid arguments are passed.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::install_package
+	 *
+	 * @dataProvider data_install_package_should_return_wp_error_with_invalid_arguments
+	 *
+	 * @param array $arguments The arguments to test against.
+	 */
+	public function test_install_package_should_return_wp_error_with_invalid_arguments( $arguments ) {
+		$this->instance->generic_strings();
+
+		$this->upgrader_skin_mock->expects( $this->never() )->method( 'feedback' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'dirlist' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'find_folder' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'is_dir' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'exists' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'mkdir' );
+
+		$actual = $this->instance->install_package( $arguments );
+
+		$this->assertWPError(
+			$actual,
+			'WP_Upgrader::install_package() did not return a WP_Error object'
+		);
+
+		$this->assertSame(
+			'bad_request',
+			$actual->get_error_code(),
+			'The WP_Error code was not "bad_request"'
+		);
+	}
+
+	/**
+	 * Data provider: Provides invalid arguments for `WP_Upgrader::install_package()`.
```

Comment:
@costdev Is this valid format for document?

Reply:
Thanks @mukeshpanchal27! This is not currently a valid format for data providers. I have pushed some updates to the PR, including to set all data provider docblocks to the current "for now" format:

```php
/**
 * Data provider.
 *
 * @return array[]
 */
```